compiler/src/jdk.graal.compiler/src/jdk/graal/compiler/hotspot/amd64/AMD64HotSpotMacroAssembler.java

```diff
@@ -140,4 +140,35 @@ public void verifyOop(Register value, Register tmp, Register tmp2, boolean compr
         illegal();
         bind(ok);
     }
+
+    @Override
+    protected final int membarOffset() {
+        // All usable chips support "locked" instructions which suffice
+        // as barriers, and are much faster than the alternative of
+        // using cpuid instruction. We use here a locked add [esp-C],0.
+        // This is conveniently otherwise a no-op except for blowing
+        // flags, and introducing a false dependency on target memory
+        // location. We can't do anything with flags, but we can avoid
+        // memory dependencies in the current method by locked-adding
+        // somewhere else on the stack. Doing [esp+C] will collide with
+        // something on stack in current method, hence we go for [esp-C].
+        // It is convenient since it is almost always in data cache, for
+        // any small C. We need to step back from SP to avoid data
+        // dependencies with other things on below SP (callee-saves, for
+        // example). Without a clear way to figure out the minimal safe
+        // distance from SP, it makes sense to step back the complete
+        // cache line, as this will also avoid possible second-order effects
```

Comment:
@shipilev we are replicating the StoreLoad barrier offset in Graal (https://bugs.openjdk.org/browse/JDK-8050147) and are wondering what are the second-order effects with locked ops against the cache line. From the Intel manual Volumn 3 Chapter 23.34 it says  

> The Pentium processor does check its cache on a read-modify-write access and, if the cache line has been modified, writes the contents back to memory before locking the bus. The P6 family processors write to their cache on a read-modify-write operation (if the access does not split across a cache line) and does not write back to system memory.

If that is the only possible second-order effect, would it be more beneficial to stay in the same cache line to avoid read miss for modern CPUs? In your benchmarking it seems [rsp-8] scores the best. 

Also, would it be possible to share your barrier benchmark http://cr.openjdk.java.net/~shade/8050147/VolatileBarrierBench.java, which is no longer available?
/cc @merykitty

Reply:
IIRC, I meant the secondary effects for sharing (probably hot) cache line that carries `(rsp+0)` with the location where we are performing the locked-as-membar access. `(rsp-8)` might be on the same cache line, which means, while there is no data dependency (all moderns CPUs are likely disambiguating stack accesses well), but we can still hit the same cache line. In multi-CPU case, that would be false sharing. In single-CPU case -- since this is a local stack, and probably not shared -- this still might have effects on how CPUs resolve the atomic access to the cache line. This might not matter for current CPUs, but doing a separate cache line looked more straight-forward.

Reply:
I think that trying to ping another cache line may have the negative effect of increasing cache pressure. While 1 cache line may not be much, it is not worth the risk of dying by 1000 cuts for no benefit (at least on current and mostly probably future architectures).